### DISCUSSION TITLE
Run a Zenoh based service into a custom Docker

### DISCUSSION THREAD

**Discussion by edgarriba** [eclipse-zenoh/roadmap#159] [Q&A] [answered]
hi! I have my own application for camera streaming which is implemented as a standole service as I want to integrate different protocols e.g Zenoh along with gRPC.

For this application, I need to have my custom dockerfile based on a nvidia image in order to install a bunch of system dependencies. I tried to run my service in that dockerfile as follows:

```bash
docker run -d \
    --name myapp \
    --privileged \
    --restart=unless-stopped \
    -p 50051:50051 \
    -p 7447:7447/tcp \
    -p 8000:8000/tcp \
    myapp
```

The server is running and no problem, however when I want to subscribe from a client in python from the outside, I'm not able to connect properly to the session. I attach a snippet similar to what I'm doing:

```python
# establish a connection to the server
config = zenoh.Config()
config.insert_json5("mode", json.dumps(args.mode))
config.insert_json5("connect/endpoints", json.dumps(args.connect))

# create a subscriber
sub = session.declare_subscriber(
    "camera/0/frame", on_frame, reliability=zenoh.Reliability.RELIABLE
)
```

Then I run as follows: `python py/client_zenoh.py -e tcp/localhost:7447 -m peer`

NOTE: the server and then connecting with the client it works outside of the dockerfile. But I need to run it inside the docker to facilitate the deployment of this app.

Any help or tips are welcomed!

**Comment by gabrik** [comment]
Hi @edgarriba, can you share the Zenoh configuration of the "server side" along with the error you are getting when trying to connect?
Also please run with `RUST_LOG=z=debug` so we can have more info

**Reply by edgarriba** [reply]
hi @gabrik , thanks for replying. Here is the log i get from my server:

```bash
[2024-09-13T09:33:36Z DEBUG zenoh::api::session] Config: Config { id: 6a4db9a6f2c98efe6878b3bfa8197327, metadata: Null, mode: Some(Peer), connect: ConnectConfig { timeout_ms: None, endpoints: Unique([]), exit_on_failure: None, retry: None }, listen: ListenConfig { timeout_ms: None, endpoints: Dependent(ModeValues { router: Some([tcp/[::]:7447]), peer: Some([tcp/[::]:0]), client: None }), exit_on_failure: None, retry: None }, scouting: ScoutingConf { timeout: None, delay: None, multicast: ScoutingMulticastConf { enabled: None, address: None, interface: None, ttl: None, autoconnect: None, listen: None }, gossip: GossipConf { enabled: None, multihop: None, autoconnect: None } }, timestamping: TimestampingConf { enabled: None, drop_future_timestamp: None }, queries_default_timeout: None, routing: RoutingConf { router: RouterRoutingConf { peers_failover_brokering: None }, peer: PeerRoutingConf { mode: None } }, aggregation: AggregationConf { subscribers: [], publishers: [] }, transport: TransportConf { unicast: TransportUnicastConf { accept_timeout: 10000, accept_pending: 100, max_sessions: 1000, max_links: 1, lowlatency: false, qos: QoSUnicastConf { enabled: true }, compression: CompressionUnicastConf { enabled: false } }, multicast: TransportMulticastConf { join_interval: Some(2500), max_sessions: Some(1000), qos: QoSMulticastConf { enabled: false }, compression: CompressionMulticastConf { enabled: false } }, link: TransportLinkConf { protocols: None, tx: LinkTxConf { sequence_number_resolution: U32, lease: 10000, keep_alive: 4, batch_size: 65535, queue: QueueConf { size: QueueSizeConf { control: 1, real_time: 1, interactive_high: 1, interactive_low: 1, data_high: 2, data: 4, data_low: 2, background: 1 }, congestion_control: CongestionControlConf { wait_before_drop: 1000 }, batching: BatchingConf { enabled: true, time_limit: 1 } }, threads: 1 }, rx: LinkRxConf { buffer_size: 65535, max_message_size: 1073741824 }, tls: TLSConf { root_ca_certificate: None, server_private_key: None, server_certificate: None, client_auth: None, client_private_key: None, client_certificate: None, server_name_verification: None, root_ca_certificate_base64: None, server_private_key_base64: None, server_certificate_base64: None, client_private_key_base64: None, client_certificate_base64: None }, unixpipe: UnixPipeConf { file_access_mask: None } }, shared_memory: ShmConf { enabled: true }, auth: AuthConf { usrpwd: UsrPwdConf { user: None, password: None, dictionary_file: None }, pubkey: PubKeyConf { public_key_pem: None, private_key_pem: None, public_key_file: None, private_key_file: None, key_size: None, known_keys_file: None } } }, adminspace: AdminSpaceConf { enabled: false, permissions: PermissionsConf { read: true, write: false } }, downsampling: [], access_control: AclConfig { enabled: false, default_permission: Deny, rules: None, subjects: None, policies: None }, plugins_loading: PluginsLoading { enabled: false, search_dirs: LibSearchDirs([Spec(LibSearchSpec { kind: CurrentExeParent, value: None }), Path("."), Path("~/.zenoh/lib"), Path("/opt/homebrew/lib"), Path("/usr/local/lib"), Path("/usr/lib")]) }, plugins: Object {} }
[2024-09-13T09:33:36Z DEBUG zenoh::net::runtime] Zenoh Rust API v1.0.0-beta.1
[2024-09-13T09:33:36Z INFO  zenoh::net::runtime] Using ZID: 6a4db9a6f2c98efe6878b3bfa8197327
[2024-09-13T09:33:36Z DEBUG zenoh::net::routing::interceptor::access_control] Access control is disabled
[2024-09-13T09:33:36Z DEBUG zenoh::net::routing::hat::p2p_peer::gossip] [Gossip] Add node (self) 6a4db9a6f2c98efe6878b3bfa8197327
[2024-09-13T09:33:36Z DEBUG zenoh::net::routing::router] New Face{0, 6a4db9a6f2c98efe6878b3bfa8197327}
[2024-09-13T09:33:36Z DEBUG zenoh::net::runtime::orchestrator] Try to add listener: tcp/[::]:0: ConnectionRetryConf { exit_on_failure: true, period_init_ms: 1000, period_max_ms: 4000, period_increase_factor: 2.0 }
[2024-09-13T09:33:36Z DEBUG zenoh::net::runtime::orchestrator] Listener added: tcp/[::]:44843
[2024-09-13T09:33:36Z INFO  zenoh::net::runtime::orchestrator] Zenoh can be reached at: tcp/172.17.0.2:44843
[2024-09-13T09:33:36Z DEBUG zenoh::net::runtime::orchestrator] UDP port bound to 224.0.0.224:7446
[2024-09-13T09:33:36Z DEBUG zenoh::net::runtime::orchestrator] Joined multicast group 224.0.0.224 on interface 172.17.0.2
[2024-09-13T09:33:36Z INFO  zenoh::net::runtime::orchestrator] zenohd listening scout messages on 224.0.0.224:7446
[2024-09-13T09:33:36Z DEBUG zenoh::net::runtime::orchestrator] UDP port bound to 172.17.0.2:56756
[2024-09-13T09:33:36Z DEBUG zenoh::net::runtime::orchestrator] Waiting for UDP datagram...
[2024-09-13T09:33:37Z DEBUG zenoh::net::routing::dispatcher::interests] Face{0, 6a4db9a6f2c98efe6878b3bfa8197327} Declare interest 2 (camera/0/frame)
[2024-09-13T09:33:37Z DEBUG zenoh::net::routing::hat::p2p_peer::interests] Propagate DeclareFinal Face{0, 6a4db9a6f2c98efe6878b3bfa8197327}:2
```

as for the config, I'm using the defaults as follows:

```rust
let zenoh_session = zenoh::open(zenoh::config::default())
        .await
        .map_err(|e| ServerError::FailedToStartZenohSession(e.to_string()))?;
```

**Reply by gabrik** [reply, marked as answer]
You are missing the actual listening locator, on your server side, please add something like `tcp/[::]:7447` in your server configuration.
Please refer to documentation on how to configure a listener: https://docs.rs/zenoh/latest/zenoh/prelude/struct.Config.html

**Reply by edgarriba** [reply]
that worked ! thanks !

**Comment by higashiyamar03** [comment]
## Deep Dive #1: Long vs Short Decomposition

### The Short Side Is the Dominant Cause of Failure

**OOF Period (2017-2024, Ensemble Rank-Average):**

| Side | Weekly Mean | Sharpe | Hit Rate | Cumulative |
|------|:---:|:---:|:---:|:---:|
| Long | +0.00376 | **+0.81** | 60.5% | +259.4% |
| Short | +0.00210 | **+0.42** | 45.9% | +82.7% |
| **Gross L/S** | +0.00293 | **+1.40** | 59.6% | +211.3% |

Both sides contributed positively. The long side was the primary driver (Sharpe 0.81 vs 0.42), but the short side was profitable too.

**Holdout Period (2025-2026, LGBM Walk-Forward):**

| Side | Weekly Mean | Sharpe | Hit Rate | Cumulative |
|------|:---:|:---:|:---:|:---:|
| Long | +0.00027 | **+0.06** | 58.0% | -1.4% |
| Short | -0.00459 | **-1.26** | 37.7% | -28.9% |
| **Gross L/S** | -0.00216 | **-1.16** | 43.5% | -14.4% |

The long side went to zero (Sharpe 0.06). **The short side went deeply negative** (Sharpe -1.26, hit rate 37.7%).

### Weekly Contribution Pattern

| Scenario | Frequency | Mean Gross Return |
|----------|:---:|:---:|
| Both sides lose | 14.5% of weeks | -0.48% |
| Only long loses | 27.5% of weeks | -0.43% |
| **Only short loses** | **47.8% of weeks** | -0.04% |
| Both sides win | 10.1% of weeks | -3.75% |

The short side is unprofitable in nearly half of all weeks (47.8%). Even when both sides win (rare, 10.1%), the gross return is still negative — suggesting the short side's losses swamp the long side's gains even in "good" weeks.

### The Short Side Crisis Is Structural

The OOF period hid the short-side fragility because the model could short stocks that:
1. Were acquired (positive pre-acquisition drift meant these were actually bad shorts)
2. Survived through 2024 (survivorship bias: the "losers" were never in the dataset)

When the model faces 2025's full universe, the short book fills with stocks the model thinks will decline — but the features it uses to identify "declining" stocks were trained on a universe where the worst outcomes were systematically excluded.

### Yearly Long/Short Sharpe

**OOF (Ensemble):**

| Year | Long Sharpe | Short Sharpe | Long Hit | Short Hit |
|------|:---:|:---:|:---:|:---:|
| 2017 | **+3.92** | -2.72 | 68% | 32% |
| 2018 | -0.25 | **+0.59** | 48% | 56% |
| 2019 | **+3.05** | -0.89 | 73% | 40% |
| 2020 | **+0.81** | -0.90 | 70% | 36% |
| 2021 | **+1.66** | +0.52 | 60% | 40% |
| 2022 | -0.18 | **+2.05** | 54% | 62% |
| 2023 | +0.90 | -0.14 | 56% | 48% |
| 2024 | -0.12 | **+1.35** | 58% | 50% |

**Holdout (LGBM WF):**

| Year | Long Sharpe | Short Sharpe | Long Hit | Short Hit |
|------|:---:|:---:|:---:|:---:|
| 2025 | -0.09 | **-1.71** | 57% | 36% |
| 2026 | **+0.65** | -0.03 | 62% | 44% |

The long/short roles historically **alternated**: some years the long side carried (2017, 2019, 2021), other years the short side carried (2018, 2022, 2024). On holdout, **both sides failed simultaneously** for the first time.

---

## Deep Dive #2: Feature Importance Shift — Zero Top-50 Overlap Across Eras

Three LGBM models were trained on non-overlapping OOF sub-periods with identical hyperparameters (n=12,000 each, 200 rounds, seed=42):

### Top-50 Feature Overlap

| Comparison | Overlap |
|------------|:---:|
| 2015-2020 ∩ 2021-2024 | **3 / 50** |
| 2015-2020 ∩ Full OOF | **2 / 50** |
| 2021-2024 ∩ Full OOF | **11 / 50** |
| All three periods | **0 / 50** |

**Zero features are in the top-50 across all three periods.** The model relies on entirely different features in each era.

### Feature Group Dominance Shift

| Group | 2015-2020 | 2021-2024 | Shift |
|-------|:---:|:---:|:---:|
| path_shape | **27.5%** | 18.1% | -9.4pp |
| ohlcv_derived | 23.4% | 19.5% | -3.9pp |
| talib | 20.8% | **34.4%** | **+13.6pp** |
| wavelet | 16.8% | 17.5% | +0.7pp |

The model migrated from path_shape dominance (Optiver-style features) in the ZIRP era to talib dominance (traditional technicals) in the post-COVID era. This is not a stable alpha source — it's a regime-chasing pattern.

### Top Features by Era

**2015-2020 (ZIRP/QE era):**
1. `log_ret_60d_minus_week_mean` — long-term momentum deviation
2. `macd_csz` — MACD cross-sectional z-score
3. `path_rv_short_long_5d_week_std` — Optiver realized vol ratio
4. `wt_approx_energy` — wavelet energy
5. `path_return_vol_60d_csz` — mid-term path volatility

**2021-2024 (post-COVID/inflation era):**
1. `atr_14_csr` — ATR cross-sectional rank
2. `mom_10_csz` — 10-day momentum z-score
3. `adosc_week_mean` — Chaikin A/D oscillator
4. `dollar_adv` — dollar volume
5. `wt_entropy_week_mean` — wavelet entropy

**Full OOF (smoothed):**
1. `wt_entropy_week_mean` — wavelet entropy (only appears in 2021-24 top-5, NOT in 2015-20)
2. `log_ret_1d_div_week_mean_safe` — short-term reversal
3. `path_close_window_start_20d_div_sector_mean_safe` — sector-relative path
4. `ret_kurt_60d` — return kurtosis (tail risk)
5. `dollar_adv` — liquidity

The Full OOF top features are dominated by the 2021-2024 period's patterns because that period has higher variance (COVID, inflation, rate hikes) and the gain-based importance rewards features that split high-variance nodes.

### What This Means

1. **There is no stable alpha factor in this feature set.** The model finds different signals in different eras. Each era's signal is a statistical artifact of that specific market regime.

2. **The 0.75 Spearman rank correlation (across all 1561 features) is misleading.** It looks reasonable because 1500+ low-importance features have stable (near-zero) ranks. The features that actually drive predictions (top-50) have zero persistence.

3. **The model is a regime detector, not a return predictor.** It learns "what worked recently" and applies it. When the regime shifts (2025), none of the learned patterns transfer.

---

## Deep Dive #3: Missing Symbols — Direct Impact Is Small, Training Impact Is Large

### Direct Holdout Impact

If WBA (Walgreens, -64% in 2024) and OKTA (-20% pre-removal) were in the 2025 universe and correctly identified as shorts:

- WBA short contribution: ~+0.025% per week when held → **~+0.9% cumulative boost**
- OKTA short contribution: ~+0.007% per week when held → **~+0.2% cumulative boost**
- **Combined direct boost: ~+1%** — negligible relative to the -16% holdout drawdown

### Training Impact (The Real Problem)

The 9 missing stocks represent **two failure modes the model never learned**:

| Failure Mode | Missing Examples | Why It Matters |
|---|---|---|
| **Secular decline** (value trap) | WBA, OKTA | The model never saw a stock slowly deteriorate and get kicked out of the index. It only saw survivors and acquisition targets (which have positive drift). |
| **Acquisition premium** | CELG, XLNX, ATVI, SGEN, SPLK | Acquired stocks had positive returns pre-deal. Being short these was actually *profitable* in training (the model learned that "stocks that look like acquisition targets go up"). |

The model's short-side logic is fundamentally broken:
- It learned that "cheap healthcare stocks with declining momentum" = **buy** (because CELG, SGEN were acquired at premium)
- In reality, "cheap healthcare stocks with declining momentum" = **short** (WBA kept falling)
- It learned that "SaaS stocks with high volatility" = **buy** (because SPLK was acquired by CSCO)
- In reality, "SaaS stocks with security problems and declining growth" = **short** (OKTA kept falling)

### The Sign Problem

The missing symbols create a **sign error** in the model's short-side predictions. The features that should identify shorts instead identify "potential acquisition targets" because 6 out of 8 removed stocks were acquired (positive returns), and only 2 were genuine losers.

**In the training data:** stocks with "removal risk" features → +returns (acquisition premium)
**In reality:** stocks with "removal risk" features → -returns (getting kicked out for poor performance)

This explains why the short side goes from Sharpe +0.42 (OOF) to -1.26 (holdout). The model is systematically long the worst stocks because its training data taught it that "troubled NASDAQ 100 constituents get acquired at a premium."

---

## Synthesis: The Failure Is Multi-Causal

| Cause | Mechanism | Severity |
|-------|-----------|:---:|
| **Short-side sign error** | Missing losers → model buys what it should short | **CRITICAL** |
| **Feature regime dependence** | Zero top-50 feature overlap across eras → no stable signal | **CRITICAL** |
| **Long-side signal decay** | Long Sharpe 0.81→0.06 → features stopped working | HIGH |
| **Amplification by turnover** | Turnover 19.7%→24.0% when signals are noise | MEDIUM |
| **Survivorship bias (direct)** | Missing 9 stocks = ~+1% holdout impact | LOW |

The two critical fixes needed:
1. **PIT universe reconstruction** — include all 105+ historical NASDAQ100 members so the model sees the full distribution of negative outcomes
2. **Feature stability filter** — features that don't appear in the top-50 across at least 2 of 3 sub-periods should be downweighted or removed; a model built on regime-specific features cannot generalize
